File: commodetto/commodetto_BufferOut.c
/*
    Copyright (C) 2016 Moddable Tech, Inc.

    This Source Code Form is subject to the terms of the Mozilla Public
    License, v. 2.0. If a copy of the MPL was not distributed with this
    file, You can obtain one at http://mozilla.org/MPL/2.0/.
*/

/*
	Commodetto Buffer in Memory
		
		implements PixelOut to write pixels to an ArrayBuffer
*/


#include "mc_xs.h"
#include "mc_misc.h"
#include "mc.xs.h"	//@@ use xsID_* in place of xsID("*")

//#define ID(symbol) (xsID_##symbol)
#define ID(symbol) (xsID(#symbol))

#define kHeader (16)

#define kOffset (0)
#define kWindowWidth (1)
#define kWindowOffset (2)
#define kRowBump (3)

void xs_BufferOut_send(xsMachine *the)
{
	int argc = xsToInteger(xsArgc);
	char *src, *dst;
	uint32_t *header;
	int offsetIn, count, offsetOut;

	if (argc >= 1)
		offsetIn = xsToInteger(xsArg(1));
	else
		offsetIn = 0;

	if (argc >= 2)
		count = xsToInteger(xsArg(2));
	else
		count = xsGetArrayBufferLength(xsArg(0)) - offsetIn;

 	if (xsIsInstanceOf(xsArg(0), xsArrayBufferPrototype))
		src = xsToArrayBuffer(xsArg(0));
	else
		src = xsGetHostData(xsArg(0));

	xsVars(1);
	xsGet(xsVar(0), xsThis, ID(buffer));
	header = xsToArrayBuffer(xsVar(0));
	dst = (char *)header;

	offsetOut = header[kOffset];

	while (count) {
		int copy = header[kWindowWidth] - header[kWindowOffset];
		if (copy > count)
			copy = count;
		count -= copy;
		header[kWindowOffset] += copy;
		memcpy(dst + offsetOut, src + offsetIn, copy);
		offsetOut += copy;
		offsetIn += copy;
		if (header[kWindowOffset] == header[kWindowWidth]) {
			header[kWindowOffset] = 0;
			offsetOut += header[kRowBump];
		}
	}

	header[kOffset] = offsetOut;
}

/*
	send(pixels, offsetIn = 0, count = pixels.byteLength - offsetIn) {
		let offsetOut = this.offset;
		let bytes = this.bytes;
		pixels = new Uint8Array(pixels);		//@@

		while (count) {
			let copy = this.windowWidth - this.windowOffset;
			if (copy > count)
				copy = count;
			count -= copy;
			this.windowOffset += copy;
			while (copy--)
				bytes[offsetOut++] = pixels[offsetIn++];
			if (this.windowOffset == this.windowWidth) {
				this.windowOffset = 0;
				offsetOut += this.rowBump;
			}
		}

		this.offset = offsetOut;
	}
*/

File: commodetto/commodetto_BufferOut.js
/*
    Copyright (C) 2016 Moddable Tech, Inc.

    This Source Code Form is subject to the terms of the Mozilla Public
    License, v. 2.0. If a copy of the MPL was not distributed with this
    file, You can obtain one at http://mozilla.org/MPL/2.0/.
*/

/*
	Commodetto Buffer in Memory
		
		implements PixelOut to write pixels to an ArrayBuffer
*/

import PixelsOut from "PixelsOut";
import Bitmap from "Bitmap";

const header = 16;			// number of bytes preceeding pixels

export default class BufferOut extends PixelsOut {
	constructor(dictionary) {
		super(dictionary);

		let format = Bitmap.Raw;
		switch (dictionary.pixelFormat) {
			case "rgb565le":
			case "rgb565be":
				this.depth = 16;
				break;

			case "g4":
				this.depth = 4;
				format = Bitmap.Gray;
				break;

			case "m1":
				this.depth = 1;
				format = Bitmap.Monochrome;
				break;

			default:
				throw new Error("unsupported buffer pixel fornat");
				break;

		}

		// initialize array buffer for use as Bitmap (ArrayBuffer + width, height, and offset properties)
		this.buffer = new ArrayBuffer(this.pixelsToBytes(this.width * this.height) + header);
		this.bits = new Bitmap(this.width, this.height, format, this.buffer, header);

		this.words = new Uint32Array(this.buffer, 0, header);
	}
	begin(x, y, width, height) {
		let words = this.words;

		words[0] = this.pixelsToBytes((y * this.width) + x) + header;		// offset
		words[1] = this.pixelsToBytes(width);								// windowWidth
		words[2] = 0;														// windowOffset
		words[3] = this.pixelsToBytes(this.width - width);					// rowBump
	}
	send(pixels, offsetIn, count) @ "xs_BufferOut_send"
	end() {
	}
	continue() {
		// empty implementation overrides PixelOut.continue which throws
	}
	pixelsToBytes(count) {
		return ((count * this.depth) + 7) >> 3;
	}
	get bitmap() {
		return this.bits;
	}
}
